### PR TITLE
Move Memento from a submodule to a plugin [Reviewer: EM]

### PR DESCRIPTION
**Pull request by tomwatson1024**
This pull request moves the Memento sproutlet from a submodule of the Sprout repo to a plugin used in the same way as greeter/houdini. This means that we'll no longer have to build thrift/cassandra and memento itself when building Sprout, unless we actually want Memento.

Lots of the new files added are just copied from the Sprout repo, but that doesn't show up in the diff.

The Memento unit tests and live tests pass.

### CODE REVIEW

File: debian/compat [hunk 1]
```diff
@@ -0,0 +1 @@
+8
```
Comment (tomwatson1024) [inline debian/compat:1]: This file was copied with no changes.

File: debian/memento-as.postinst [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    abort-upgrade|abort-remove|abort-deconfigure)
+    ;;
+
+    *)
+        echo "postinst called with unknown argument \`$1'" >&2
+        exit 1
+    ;;
+esac
+
+exit 0
```
Comment (tomwatson1024) [inline debian/memento-as.postinst:47]: This file was copied without changes.

File: debian/memento-as.postrm [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+#
+# see: dh_installdeb(1)
+
+set -e
+
+# Restart sprout so it unloads the plug-in that has been removed.
+sudo service sprout stop || /bin/true
+
+exit 0
```
Comment (tomwatson1024) [inline debian/memento-as.postrm:20]: This file was copied without changes.

File: debian/memento-as.prerm [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    failed-upgrade)
+    ;;
+
+    *)
+        echo "prerm called with unknown argument \`$1'" >&2
+        exit 1
+    ;;
+esac
+
+exit 0
```
Comment (tomwatson1024) [inline debian/memento-as.prerm:41]: This file was copied without changes.

File: debian/memento-as.triggers [hunk 5]
```diff
@@ -0,0 +1 @@
+activate clearwater-snmp-alarm-agent
```
Comment (tomwatson1024) [inline debian/memento-as.triggers:1]: This file was copied without changes.

File: debian/rules [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+override_dh_auto_test:
+	mkdir debian/tmp
+
+# Don't start - monit does that for us.
+override_dh_installinit:
+	dh_installinit --no-start -u"defaults 60 40"
+
+override_dh_shlibdeps:
+override_dh_strip:
+	dh_strip -pmemento-as --dbg-package=memento-as-dbg
```
Comment (tomwatson1024) [inline debian/rules:34]: This file was copied without changes, except perhaps the package name.

File: memento-as.root/usr/share/clearwater/infrastructure/alarms/memento_as_alarms.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+                    "details": "While this condition persists, new call list records will not be written and will be lost. The Application Server will periodically attempt to reconnect.",
+                    "description": "Memento: Application Server Cassandra communication error",
+                    "cause": "Memento Application Server is unable to contact the local Cassandra.",
+                    "effect": "While this condition persists, new call list records will not be written and will be lost.",
+                    "action": "Monitor for the alarm to clear and confirm the system is operating normally. Determine if the local Cassandra process has failed. If it has then make sure it returns to service. If the alarm doesn't clear then contact your support representative."
+                }
+            ]
+        }
+    ]
+}
```
Comment (tomwatson1024) [inline memento-as.root/usr/share/clearwater/infrastructure/alarms/memento_as_alarms.json:27]: This file was copied without changes.

File: mk/cassandra.mk [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+cassandra_clean:
+	rm -rf ${CASSANDRA_GEN_DIR}
+
+cassandra_distclean:
+	rm -rf ${CASSANDRA_BUILD_DIR}
+
+-include ${D_PATHS}
+
+.PHONY: cassandra cassandra_test cassandra_clean cassandra_distclean
```
Comment (tomwatson1024) [inline mk/cassandra.mk:53]: This file was copied without changes.

File: mk/thrift.mk [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+thrift_clean: ${THRIFT_MAKEFILE}
+	${MAKE} -j1 -C ${THRIFT_DIR} clean
+
+thrift_distclean:
+	# The following doesn't seem to work, so use git clean instead
+	# ${MAKE} -C ${THRIFT_DIR} distclean
+	cd ${THRIFT_DIR} && git clean -f -X -d
+
+.PHONY: thrift thrift_test thrift_clean thrift_distclean
```
Comment (tomwatson1024) [inline mk/thrift.mk:31]: This file was copied without changes. Note the known issue [#2078](https://github.com/Metaswitch/clearwater-issues/issues/2078) with building thrift.

File: memento-as.root/usr/share/clearwater/sprout/plugin_conf.d/memento-as.plugin_conf [hunk 10]
```diff
@@ -0,0 +1,51 @@
+#!/bin/sh
+
+# @file memento-as.plugin_conf
+#
+# Copyright (C) 2017  Metaswitch Networks Ltd
```
Comment (eleanor-merry) [inline memento-as.root/usr/share/clearwater/sprout/plugin_conf.d/memento-as.plugin_conf]: Should now be 2018

File: src/mementoasplugin.cpp [hunk 11]
```diff
@@ -0,0 +1,325 @@
+/**
+ * @file mementoasplugin.cpp  Plug-in wrapper for the Memento Sproutlet.
+ *
+ * Copyright (C) Metaswitch Networks 2016
```
Comment (eleanor-merry) [inline src/mementoasplugin.cpp]: 2018 (it's quite different from the old file)

File: src/mementoasplugin.cpp [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    opt = opt_it->second;
+    TRC_INFO("%s memento-as option '%s': '%s'",
+             required_opt ? "Set" : "Overwrote",
+             opt_name.c_str(),
+             opt.c_str());
+  }
+  else if (required_opt)
+  {
+    TRC_STATUS("Required memento-as option '%s' not set. Disabling memento-as.",
+              opt_name.c_str());
```
Comment (eleanor-merry) [inline src/mementoasplugin.cpp]: alignment

File: src/mementoasplugin.cpp [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+    opt = as_int;
+    TRC_INFO("%s memento-as option '%s': %d",
+             required_opt ? "Set" : "Overwrote",
+             opt_name.c_str(),
+             opt);
+  }
+  else if (required_opt)
+  {
+    TRC_STATUS("Required memento-as option '%s' not set. Disabling memento-as.",
+              opt_name.c_str());
```
Comment (eleanor-merry) [inline src/mementoasplugin.cpp]: alignment

File: src/mementoasplugin.cpp [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+    SNMP::SuccessFailCountByRequestTypeTable* incoming_sip_transactions_tbl = SNMP::SuccessFailCountByRequestTypeTable::create("memento_as_incoming_sip_transactions",
+                                                                                                                               "1.2.826.0.1.1578918.9.8.1.4");
+    SNMP::SuccessFailCountByRequestTypeTable* outgoing_sip_transactions_tbl = SNMP::SuccessFailCountByRequestTypeTable::create("memento_as_outgoing_sip_transactions",
+                                                                                                                               "1.2.826.0.1.1578918.9.8.1.5");
+    _cass_comm_monitor = new CommunicationMonitor(new Alarm(alarm_manager,
+                                                            "memento",
+                                                            AlarmDef::MEMENTO_AS_CASSANDRA_COMM_ERROR,
+                                                            AlarmDef::CRITICAL),
+                                                  "Memento",
+                                                  "Memcached");
```
Comment (eleanor-merry) [inline src/mementoasplugin.cpp]: This should probably be Cassandra

File: ut/coverage-not-yet [hunk 15]
```diff
@@ -0,0 +1,7 @@
+include/call_list_store_processor.h
```
Comment (eleanor-merry) [inline ut/coverage-not-yet]: This is a lot of uncovered files! What kind of coverage is there?

File: src/mementoasplugin.cpp [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+
+/// Unloads the Memento plug-in.
+void MementoPlugin::unload()
+{
+  delete _memento_sproutlet;
+  delete _memento;
+  delete _cass_resolver;
+  delete _call_list_store;
+  delete _cass_comm_monitor;
+}
```
Comment (eleanor-merry) [inline src/mementoasplugin.cpp:325]: You've moved all the uts from src/ut to ut. I assume this is to match up with Houdini?

**Review by eleanor-merry — APPROVED** [review]
LGTM - some minor comments and one question about coverage.

File: src/mementoasplugin.cpp (hunk 16 above)
Comment (tomwatson1024) [reply src/mementoasplugin.cpp:325]: Yes, and also ensures compatibility with gcov/build-infra.

File: ut/coverage-not-yet (hunk 15 above)
Comment (tomwatson1024) [reply ut/coverage-not-yet]: Coverage is pretty poor in this area, although no poorer as a result of these changes. Worth me adding some UTs?
```
------------------------------------------------------------------------------
                           GCC Code Coverage Report
Directory: /home/ubuntu/sprout/plugins/memento-as
------------------------------------------------------------------------------
File                                       Lines    Exec  Cover   Missing
------------------------------------------------------------------------------
src/mementoappserver.cpp                     195     195   100%
src/httpnotifier.cpp                          31      16    51%   48,51,54,57-61,63-65,67-68,73-74
modules/memento-common/src/mementosaslogger.cpp
                                               3       1    33%   14,19
src/call_list_store_processor.cpp            101      16    15%   51,63,65-71,73-74,77,81-85,88,91-92,99,101,104-105,107,111,113,115,117,121,123,125,129,131-132,137-138,142-143,145-146,149-150,156,167-168,171-172,174,181,186,190,195,197,199,202,207,209-210,212,215,217,220-221,223,229-231,234-235,237,244,246,248,250,253,255-259,261,267-268,271
modules/memento-common/src/call_list_store.cpp
                                             167       7     4%   42,44,47,50,53,73,76,78,80,82,84,86,88,98,101,107-111,117,119,129,137-138,140-141,143,147,149,152-157,165-169,172-173,176-177,183,186-187,190,193,197,199-200,204-205,208,213,220-222,224-225,227,231,234-236,240,246,248-249,255-256,258,266-268,270,272,275-279,281,283,285,288-292,295,298,302,304-305,309-310,312,314-315,318,320,327,333-334,336-337,339,342,344,347-350,358-360,363-371,374,376,379-380,383,386,390,392-393,397-398,401,405,414,423,425-426,428-429,434,438,440,442,445,447-448,453,460-462,464-465
modules/memento-common/include/call_list_store.h
                                               1       0     0%   21
include/call_list_store_processor.h            3       0     0%   57,68,72
------------------------------------------------------------------------------
TOTAL                                        501     235    46%
------------------------------------------------------------------------------
```

**Pull request merged**